### PR TITLE
Pedram/calendar finished

### PR DESCRIPTION
**Pull request by namiranianp**
I decided to just manually create the ICS file using strings because it is equally/less difficult than using the functions provided by javascript libraries. 
The main logic is in calendar.js but is called by a function in planner.js, this is because it seems simpler to get the desired plan in planner.js than in calendar.js

### CODE REVIEW

File: client/assets/calendar.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
+import { saveAs } from 'file-saver'
+import Moment from 'moment'
+import {
+  extendMoment
+} from 'moment-range'
+
+export async function downloadCalendar({ plan }) {
+	const FileSaver = require('file-saver')
```
Comment (nperez0111) [inline client/assets/calendar.js]: This guy is a little outdated, either use the import statement you have above (and if that doesn't work as it may not destructure to `saveAs` properly, just do `import FileSaver from 'file-saver';`)

File: client/assets/calendar.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	const moment = extendMoment(Moment)
```
Comment (nperez0111) [inline client/assets/calendar.js]: This is constant so it should be hoisted up outside this function so it doesn't get created every time we download (not that it is that costly just out of good practice hoist all constants above the function so the function can do its work.

File: client/assets/calendar.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	const ics_day_list = ['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA'] //days of the week for ICS standard
```
Comment (nperez0111) [inline client/assets/calendar.js]: Same for this guy

File: client/assets/calendar.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	const semester_year = (plan.courses[0].dates[0].slice(0, 2) < moment().month()) ? moment().year() + 1 : moment().year()
+
+	//required iCalendar info
+	//calendar_string is an array because what we use to print requires an array
+	let calendar_string = ['BEGIN:VCALENDAR\nPRODID:-//Nick and Pedram, Squaaaad//Course Calendar//EN\nVERSION:2.0\nCALSCALE:GREGORIAN']
```
Comment (nperez0111) [inline client/assets/calendar.js]: First, you define it to be an array then start appending to it making it into a string. Just make it a string.

Also to not have to do all this constant appending look into [template strings in JavaScript](http://lmgtfy.com/?q=template+strings)

File: client/assets/calendar.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+				course_dows.push(5)
+			break;
+				case 'S':
+				course_dows.push(6)
+			break;
+				case 'U':
+				course_dows.push(0)
+			break;
+		}
+		})
```
Comment (nperez0111) [inline client/assets/calendar.js]: This can be rewritten to be:
``` JavaScript
const course_dows = plan_course.days.map((arr, day) => {
     switch (day) {
       case "M":
         return arr.concat(1);
       case "T":
         return arr.concat(2);
       case "W":
         return arr.concat(3);
       case "R":
         return arr.concat(4);
       case "F":
         return arr.concat(5);
       case "S":
         return arr.concat(6);
       case "U":
         return arr.concat(0);
       default:
         return arr;
     }
   });
```

File: client/assets/calendar.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+		//there is an issue with the first day of the semester being later than when the first day of this class is
+		while(!course_dows.includes(start_moment.day())) {
+		start_moment.add(1, 'day')
+		}
+
+		let right_now = (moment().year()).toString() //used to show when this was made in correct ICS format
```
Comment (nperez0111) [inline client/assets/calendar.js]: If you want moment to output into a format use the [.format()](https://momentjs.com/docs/#/displaying/format/) function I believe this may even be the format you are looking for here: [ISO](https://momentjs.com/docs/#/displaying/as-iso-string/)

This code is way too complicated for what it is actually doing

File: client/assets/calendar.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		right_now += (moment().date().toString().length == 2) ? moment().date() : '0' + moment().date()
+		right_now += 'T'
+		right_now += (moment().hours().toString().length == 2) ? moment().hours() : '0' + moment().hours()
+		right_now += (moment().minutes().toString().length == 2) ? moment().minutes() : '0' + moment().minutes()
+		right_now += '00Z'
+
+		//set up timezone info for this event
+		calendar_string += 'BEGIN:VEVENT\nDTSTART;TZID=America/Los_Angeles:'
+		calendar_string += start_moment.year()
+		calendar_string += ((start_moment.month() + 1).toString().length == 2) ? start_moment.month() + 1 : '0' + (start_moment.month() + 1)
```
Comment (nperez0111) [inline client/assets/calendar.js]: The same applies here: use the library to output in the format you want. no need to do all this.

File: client/assets/calendar.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		calendar_string += '00'
+		//set what days this course is on, aka the repititionss
+		calendar_string += '\nRRULE:FREQ=WEEKLY;WKST=SU;UNTIL='
+		calendar_string += end_moment.year() 
+		calendar_string += ((end_moment.month() + 1).toString().length == 2) ? end_moment.month() + 1 : '0' + (end_moment.month() + 1)
+		calendar_string += (end_moment.date().toString().length == 2) ? end_moment.date() : '0' + end_moment.date()
+		calendar_string += 'T105959Z;BYDAY='
+		course_dows.forEach(function(dow) {
+			calendar_string += ics_day_list[dow] + ','
+		})
```
Comment (nperez0111) [inline client/assets/calendar.js]: Instead do: 
``` JavaScript
calendar_string += course_dows.map((dow)=>ics_day_list[dow]).join(',');
```

File: client/assets/calendar.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		calendar_string = calendar_string.slice(0, -1) //remove the extra comma at the end
```
Comment (nperez0111) [inline client/assets/calendar.js]: Won't need this if you apply above

File: client/assets/calendar.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		//other event information
+		calendar_string += '\nDTSTAMP:' + right_now + '\nUID:' + plan_course.title + '@usf.nickthesick.com\nCREATED:' + right_now
+		calendar_string += '\nLAST-MODIFIED:' + right_now + '\nLOCATION:' + plan_course.loc + '\nSEQUENCE:0\nSTATUS:CONFIRMED\nSUMMARY:'
+		calendar_string += plan_course.title + '\nEND:VEVENT\n'
```
Comment (nperez0111) [inline client/assets/calendar.js]: Template Stings are your friend they are even multiline so you can actually what the format to the file is

File: client/assets/calendar.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	})
+	calendar_string += 'END:VCALENDAR\n'
+
+	var blob = new Blob([calendar_string], {type: "text/plain;charset=utf-8"});
+	FileSaver.saveAs(blob, plan.title + ".ics");
```
Comment (nperez0111) [inline client/assets/calendar.js]: Does this ever fail at all? If it does this function should swallow the error and return false if it failed and true if it succeeded so I can tell the user later if it was successful or not

File: client/assets/calendar.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	calendar_string += '\nX-WR-TIMEZONE:America/Los_Angeles\nBEGIN:VTIMEZONE\nTZID:America/Los_Angeles'
+	calendar_string += '\nX-LIC-LOCATION:America/Los_Angeles\nBEGIN:DAYLIGHT\nTZOFFSETFROM:-0800\nTZOFFSETTO:-0700\nTZNAME:PDT'
+	calendar_string += '\nDTSTART:19700308T020000\nRRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU\nEND:DAYLIGHT\nBEGIN:STANDARD\nTZOFFSETFROM:-0700'
+	calendar_string += '\nTZOFFSETTO:-0800\nTZNAME:PST\nDTSTART:19701101T020000\nRRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU\nEND:STANDARD'
+	calendar_string += '\nEND:VTIMEZONE\n'
+
+	plan.courses.forEach(function(plan_course) {
+		//need to use moments to deal with the bug with the start day of the semester not being first day of the course
+		//start_moment = the starting date of the course and time the course starts each day
+		let start_moment = moment(plan_course.dates[0] + '/' + semester_year)
```
Comment (nperez0111) [inline client/assets/calendar.js]: Moment can parse from a specific format that you specify that would help to account for 24 hr time and not have to slice the string: [String + Format Constructor](https://momentjs.com/docs/#/parsing/string-format/)

File: client/assets/calendar.js (hunk 5 above)
Comment (nperez0111) [reply client/assets/calendar.js]: Also, I have no clue what `course_dows` is even supposed to mean I think: `plan_day_numbers` may be better of a name but I honestly have no clue what this is being used for.

File: client/package.json [hunk 13]
```diff
@@ -23,12 +23,14 @@
...
     "cookie-parser": "^1.4.3",
     "express-jwt": "^5.3.1",
+    "file-saver": "^2.0.1",
     "ics": "^2.13.2",
     "js-search": "^1.4.2",
     "local-storage": "^1.4.2",
     "moment": "^2.22.2",
     "moment-range": "^4.0.2",
     "nano": "^8.0.1",
+    "net": "^1.0.2",
```
Comment (nperez0111) [inline client/package.json]: I don't think you actually ended up using this `net` library if not please remove it.

File: client/store/planner.js [hunk 14]
```diff
@@ -106,46 +105,42 @@ export const actions = {
...
+  async loadSettings({
+    commit,
+    rootState: { auth: { loggedIn, user } }
+  }, { $api }) {
+    const savedPlans = ls('plans'),
+    savedPlan = ls('plan')
     if (loggedIn) {
       const { plans, plan } = await $api.auth.getPlans()
-      await commit('setPlans', plans || savedPlans || DEFAULT_PLANS)
+      await commit('setPlans', plans || savedPlan || DEFAULT_PLANS)
```
Comment (nperez0111) [inline client/store/planner.js]: This bit looks like it was correct, not sure why you changed it?

File: client/store/planner.js [hunk 15]
```diff
@@ -106,46 +105,42 @@ export const actions = {
...
       await commit('setCurPlan', plan || savedPlan || 0)
     } else {
+      const savedPlans = ls('plans')
```
Comment (nperez0111) [inline client/store/planner.js]: Aren't you just redeclaring the `savedPlans` again with what should be the same value? Why is this modified?

File: client/assets/calendar.js (hunk 1 above)
Comment (namiranianp) [reply client/assets/calendar.js]: Now I'm using "import FileSaver from 'file-saver';" instead of that

File: client/store/planner.js (hunk 15 above)
Comment (namiranianp) [reply client/store/planner.js]: you're right, but I don't remember doing this and I don't see the purpose of me doing it. I'll change it back

File: client/store/planner.js (hunk 14 above)
Comment (namiranianp) [reply client/store/planner.js]: I am also not sure why this was changed, I'll change it back. It was probably done when I did ctrl + f to change another variable and this got caught in it.

Comment (namiranianp) [reply client/store/planner.js]: I also do not know why I changed it, probably accidentally when changing something else named "plans." I changed it back

File: client/store/planner.js (hunk 15 above)
Comment (namiranianp) [reply client/store/planner.js]: I do not know why this is here either but without it, the download won't work. I get this error: 
Screen Shot 2019-05-02 at 1 31 06 PM

File: client/assets/calendar.js (hunk 11 above)
Comment (namiranianp) [reply client/assets/calendar.js]: the only way it "fails" is by making an unreadable ICS file, which will only become known when the user tries to open it. Regardless, I surrounded it with a try catch

File: client/assets/calendar.js (hunk 5 above)
Comment (namiranianp) [reply client/assets/calendar.js]: This code breaks my computer when run, so I use this function but have it return the number instead of returning arr.concat. This method works and doesn't cause my computer to freak out.

File: client/assets/calendar.js (hunk 2 above)
Comment (namiranianp) [reply client/assets/calendar.js]: Did you mean for this to be about downloadPlanner in planner.js? Because this is the same function as the comment below

**Comment by nperez0111** [comment]
I went ahead and cleaned up the code if you want you can look at it so you can see how I went about it.
